File: src/leetCode/c++/Solution703.cpp
//Heap KthLargest(): O(nlogk) O(k) add(): O(logk) O(k), n = nums.size()
class KthLargest {//Solution703
public:
    priority_queue<int, vector<int>, greater<int>> pq;//小到大
    int k;

    KthLargest(int k, vector<int>& nums) {
        this->k = k;
        for(int num : nums)
            add(num);
    }

    int add(int val) {
        pq.push(val);
        if(pq.size() > k) pq.pop();
        return pq.top();
    }
};

File: java/Solution703.java
package leetCode.java;

import java.util.*;

//Heap KthLargest(): O(nlogk) O(k) add(): O(logk) O(k), n = nums.length
class KthLargest {//Solution703
    PriorityQueue<Integer> pq; // min heap, 小到大
    int maxSize;

    public KthLargest(int k, int[] nums) {
        maxSize = k;
        pq = new PriorityQueue<>();
        for (int num : nums) {
            add(num);
        }
    }
    
    public int add(int val) {
        //returns the kth largest element
        pq.offer(val);
        if (pq.size() > maxSize) {
            pq.poll();
        }
        return pq.peek();
    }
}